rfc/accepted/0021-stream-processors.md

```diff
@@ -0,0 +1,205 @@
+───package data──▶│framing package processor│
+                  │                         ├──stream data──▶
+                  └─────────────────────────┘
+```
+
+Framing packet processors take package data and return new package data, but with framing information, their output can either be treated as streaming or packet data. These packet processors are required for transitioning from one model to another, but they only work in the outbound direction—they have an analougus 'framing' stream processor on the inbound side. An example here would be the `separate` processor delemiting evens by newlines or the `length-prefix` processor.
+
+### stream processors
+
+Stream processors have no notion of events. They operate on a stream of bytes and produce a transformed stream of bytes.
+
+#### 'pure' stream processors
+
+```
+                   ┌────────────────────────┐
+────stream data───▶│ pure stream processor  │────stream data───▶
+                   └────────────────────────┘
+                   ┌────────────────────────┐
+◀───stream data────│ pure stream processor  │◀───stream data────
+                   └────────────────────────┘
+```
+'pure' stream processors take and produce stream data. Any boundaries between event data in the stream are ignored and likely change as data is passed through the processor. An example would be streaming zx2 compression of events into an S3 object.
+
+
+#### 'framing' stream processors
+
+```
+                   ┌─────────────────────────┐
+◀───package data───┤                         │
+                   │framing stream processor │◀──stream data───
+◀──streaming data──┤                         │
+                   └─────────────────────────┘
+```
+
+Framing stream processors take streaming data and subdivide it into framed package data. The result can be treated as packet data or subdivided as stream data. This stream processor is required to transition from one model to another. Still, they only work in the inbound direction - they have an analogous 'framing' packet processor on the outbound side. An example here would be the `separate` processor splitting a stream into evens delimited by newlines, or the `length-prefix` processor that reads the length of the framed event and extracts as many bytes.
+
+### changes to codecs
+
+To enable codecs to produce hygenic errors and avoid misconfiguration codecs need to be able to declare if they produce or accept stream or package data. Any codecs, impemented as of the writing of this RFC, purely act on packet data.
+
+### changes to connectors
+
+Just  as codecs, connectors need to be able to declare if they work with stream or package data, below a survey of a supset of connectors implemented today. As mentioned above, structured connectors are not affected.
+
+| Connector |  type  |
+|-----------|--------|
+| tcp       | stream |
+| udp       | packet |
+| stdio     | stream |
+| sockets   | stream |
+| kafka     | packet |
+| ws        | packet |
```

Comment:
Couldn't we theoretically handle every connector input either as stream or packet? It just doesn't make sense for some (e.g. treating raw file input chunks as packets). Just wanted to make the case that for most connectors, it might be worthwile to be able to treat them as both packet- and stream-based?

Reply:
As outlined above, we technically could but we really shouldn't. from the packet connectors in there, websockets is the only one that wouldn't create total misery with something like chunking since it's a single stream reliable packet delivery